suite/tests/api/ir_aarch64.c

```diff
@@ -4532,6 +4532,88 @@ test_opnd(void *dc)
      */
 }
 
+static uint
+test_mov_instr_addr_encoding(void *dc, instr_t *instr, uint opcode, uint target_off,
+                             uint right_shift_amt, uint mask)
+{
+    ASSERT(opcode == OP_movz || opcode == OP_movk);
+    instr_t *decin;
+    byte *pc;
+
+    ASSERT(instr_get_opcode(instr) == opcode);
+    ASSERT(instr_is_encoding_possible(instr));
+
+    pc = instr_encode(dc, instr, buf);
+    decin = instr_create(dc);
+    decode(dc, buf, decin);
+
+    ASSERT(instr_get_opcode(decin) == opcode);
+
+    uint src_op = opcode == OP_movz ? 0 : 1;
+    uint expected_imm = ((ptr_int_t)buf + target_off >> right_shift_amt) & mask;
+    ASSERT(opnd_get_immed_int(instr_get_src(decin, src_op)) == expected_imm);
+
+    instr_destroy(dc, instr);
+    instr_destroy(dc, decin);
+}
+
+static void
+test_mov_instr_addr(void *dc)
+{
+    instr_t *label_instr = instr_create_0dst_0src(dc, OP_LABEL);
+    instr_set_note(label_instr, (void *)(ptr_int_t)0x100);
```

Comment:
(Hmm, I guess instr_encode() does publicly talk about setting the note fields to offsets.)

Reply:
It does. But it doesn't really prevent collisions in the note field, especially in multiple pre-encode use-cases.